### PR TITLE
make subset and canonicalise deal with individual parents

### PR DESCRIPTION
**Pull request by petrelharp**
As discussed in  #1199, the plan is for 
1. subset to not reorder individuals
2. canonicalise to sort on (num descendants, first referring node)

WIP.

### CODE REVIEW

**Comment by petrelharp** [comment]
This would close #1223.

**Comment by petrelharp** [comment]
Well, this is bothersome: I get some VCF-related errors like so:
```
____________________________________________________________________________ TestRecordsEqual.test_simple_infinite_sites_random_ploidy ____________________________________________________________________________
[gw1] linux -- Python 3.8.6 /usr/bin/python3

self = <tests.test_vcf.TestRecordsEqual object at 0x7ff680603ac0>

    def test_simple_infinite_sites_random_ploidy(self):
        ts = msprime.simulate(10, mutation_rate=1, random_seed=2)
        ts = tsutil.insert_random_ploidy_individuals(ts)
        assert ts.num_sites > 2
>       self.verify(ts)

tests/test_vcf.py:204: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_vcf.py:401: in verify
    pysam_records = list(bcf_file)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   OSError: unable to parse next record

pysam/libcbcf.pyx:4108: OSError
---------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------
[E::vcf_parse_format] Number of columns at 1:0 does not match the number of samples (4 vs 5)
============================================================================================= short test summary info =============================================================================================
FAILED tests/test_vcf.py::TestRecordsEqual::test_simple_infinite_sites_random_ploidy - OSError: unable to parse next record
```
that are apparently triggered by having individual parents - removing [these two lines from `tsutil.insert_random_ploidy_individuals`](https://github.com/tskit-dev/tskit/pull/1226/files#diff-34750e7650ea81143f77688e563dc64f9b86edccd1d765fa93cd87677d6925f8R240) makes the errors go away.

**Comment by petrelharp** [comment]
I think this is ready for review - @benjeffery?

File: c/tskit/tables.h
Comment (jeromekelleher) [inline c/tskit/tables.h:2770]: We should clarify somewhere that its *just* the individuals referred to by nodes that are retained. We don't keep the rest of the pedigree reachable from those individuals (right?)

**Review by jeromekelleher — COMMENTED** [review]
Looks great, thanks for working all this out @petrelharp! Some minor comments.

File: c/tskit/tables.c
Comment (petrelharp) [reply c/tskit/tables.c:9959]: hah, true

File: c/tskit/tables.c
Comment (petrelharp) [reply c/tskit/tables.c:9984]: Ah, so we are. You're right, it's OK here, but I'll add a comment to point this out.

File: c/tskit/tables.h
Comment (petrelharp) [reply c/tskit/tables.h:2770]: Yeah, that'd be more like `simplify`.

File: c/tskit/tables.c
Comment (petrelharp) [reply c/tskit/tables.c:5167]: Done in fd2e8b4230b862af7787b335fcb0aded922ff6a0! It was pretty straightforward, although maybe a bit confusing that there's these two outputs, each one only getting used in one place.

Comment (petrelharp) [reply c/tskit/tables.c:5169]: And in 853a5edb2d666467548fe0bffc3900357aaf4809 I've made doing the actual counting optional, to speed up the usual case a bit - don't know if this is good practice?

**Comment by petrelharp** [comment]
Codecov is complaining, but not for a good reason, I think.

File: c/tskit/tables.c
Comment (jeromekelleher) [inline c/tskit/tables.c:4986]: This is nitpicking, but we'd normally put any optional arguments last - you'd never pass in NULL for traversal_order, right?

File: c/tskit/tables.c
Comment (jeromekelleher) [inline c/tskit/tables.c:5079]: Might as well use ``sizeof(*new_id_map))`` here for consistency with mallocs above

File: c/tskit/tables.c
Comment (jeromekelleher) [inline c/tskit/tables.c:4985]: Hmm, it's not just counting descendents though. How about ``tsk_individual_table_topological_sort``? This returns a topological sorting of the individuals and optionally a count of their descendents.

File: c/tskit/tables.c
Comment (jeromekelleher) [reply c/tskit/tables.c:5169]: Looks great. Re the perf, it's probably quite marginal if it's actually worth making it optional. On the one hand, by adding some conditionals into an inner loop, you're hurting performance, making the compilers job harder, etc. On the other, we're avoiding touching a O(n) lump of memory, so keeping the cache cleaner.

 I'd be surprised if the difference was measurable on anything other than very large tables, but the code clarity of not having to allocate an array you don't use as well as keeping the memory footprint down is good, so I think you've done the right thing.

**Review by jeromekelleher — APPROVED** [review]
Love it, looks great. A couple of minor clarification points on naming things, but good to merge after squash I think.

File: c/tskit/tables.c
Comment (benjeffery) [reply c/tskit/tables.c:5169]: This looks good. The impact of the conditional will be minimal as the CPUs branch prediction will always be correct with a constant flag.

File: python/tests/test_tables.py
Comment (benjeffery) [inline python/tests/test_tables.py:3626]: Record individuals is now the default since #1238

File: python/tskit/metadata.py
Comment (benjeffery) [inline python/tskit/metadata.py:531]: Ooo, nice catch. I should have done a message test for that error.

**Review by benjeffery — APPROVED** [review]
Brilliant stuff! Once JKs comments are resolved we can merge and release!

File: python/tests/test_tables.py
Comment (petrelharp) [reply python/tests/test_tables.py:3626]: Thought so, but I didn't see it... oh, I need to rebase.

**Comment by petrelharp** [comment]
There's some kind of bug; I'm tracking it down.

File: python/tests/tsutil.py
Comment (petrelharp) [inline python/tests/tsutil.py:788]: doh!!!

File: python/tests/test_tables.py
Comment (petrelharp) [inline python/tests/test_tables.py:3764]: This is checking both the tskit union and py_union for consistency; we don't need to do that now that they both pass, I guess - but I did this to find the error in py_union.

**Comment by petrelharp** [comment]
I did a much more thorough test for consistency of union; it took longer than expected, but it's a good test. Hopefully everything passes, now?

File: python/tests/test_tables.py
Comment (benjeffery) [reply python/tests/test_tables.py:3764]: These are great and close #1243

File: python/tests/tsutil.py
Comment (benjeffery) [reply python/tests/tsutil.py:788]: Gah, that looks like my bad, sorry! I thought this would have been caught by the tests. Also a good argument for allowing assignment in the tables api.

**Comment by benjeffery** [comment]
All looks good, will merge and release tomorrow!

File: python/tests/test_tables.py
Comment (petrelharp) [reply python/tests/test_tables.py:3764]: oh sweet

Comment (petrelharp) [reply python/tests/test_tables.py]: To really close that one we should tests mutations and sites too; I've done this and updated.

**Review by jeromekelleher — APPROVED** [review]
LGTM, let's merge and ship!

**Pull request merged**
